### ISSUE TITLE
Add support for Dolphin/EnOcean PTM 215B

### ISSUE THREAD

**Issue by pfink** [sputnikdev/eclipse-smarthome-bluetooth-binding#73]
I have some battery-less wall switches based on [Dolphin/EnOcean PTM 215B](https://www.enocean.com/en/enocean-modules-24ghz-ble/details/ptm-215b/) that I'd like to integrate (technical manual can be found [here](https://www.enocean.com/enocean-modules-24ghz-ble/details/ptm-215b/user-manual-pdf/)).

My setup:

- Raspberry Pi 3B
- openHAB 2.3
- Raspbian Stretch
- Bluez 5.43
- Generic BLE 4.0 USB adapter

Some information that I found out so far:

- Bluetooth version unclear for me (but maybe not relevant)? I guess BLE 4.0 or BLE 4.2.
- Currently, the wall switches don't get discovered at all. Neither via this binding nor via `bluetoothctl` (also not when trying to do radio-based commissioning according the docs, chapter 5.3). Other BLE devices are found without any issues.
- According to the docs, the wall switches send data telegrams on each press/release event on all three advertising channels. The data channels aren't used at all.
- Authentication should happen via AES-128 CCM (counter with CBC-MAC), the security key can be read manually via QR Code on the switch, via NFC or with radio-based/BLE commissioning. But radio-based/BLE commissioning is maybe disabled, depending on the manufacturer where the end product was bought. EnOcean's default is enabled. If disabled, it can be reenabled via NFC. Anyhow, for a start I'd prefer to manually read the security key via QR code and put it statically to a textual configuration, as it seems to be the simplest approach to start with.

@vkolotov: I'm a ESH/openHAB developer, but I'm quite new to the BLE protocol. If you could give me some hints and point me to the right direction, I may could come up with a PR.

**Comment by vkolotov** [comment]
Hi @pfink, thanks for that.

* BT version is not important.
* Since your switches are battery-less, this means that they only advertise data when you actuate them. So I'd suggest to try to continuously switch it on and off until you see in the OH inbox. Otherwise, you could try to add them manually. Let me know how it goes, if not working for you I'll try to explain how to do it manually.
* Re " advertising channels" that's good, this means that you don't have to be connected to them. This is expected, good design.
* Not sure why would you need any authentication though as you've mentioned that they advertise data without establishing any secure connection. Try what I suggested above. The binding supports some basic authentication methods, e.g. auth via pin code, but that should not be required in your instance.

I'm happy to help you, let's start with getting them into inbox first.

**Comment by pfink** [comment]
OMG... Before my first post, I troubleshooted two days to get the buttons discovered including that I tried to manually add them, everything unsuccessful. The solution was as dumb as easy: I did not press the button often enough :( As you suggested, I flooded openHAB now with very often button presses (> 10) and now they're discovered..... Thank you very much!

Via online channel and by putting online timeout to 1 sec, I'm now already able to detect whether a button was pressed at all. Next step would be to distinguish which button was pressed. My preferred solution would be to implement/offer a [`system.rawrocker` channel](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/DefaultSystemChannelTypeProvider.java#L123).

Of course, I'm also open to simpler/easier solutions for a start :)

**Comment by vkolotov** [comment]
Good stuff! Very interesting to see battery-less devices to work with the binding :) What channels have been discovered? Can you post a screenshot?

**Comment by pfink** [comment]
Sure! This is a complete list:

- bluetooth:ble:xxxx:online
- bluetooth:ble:xxxx:connected
- bluetooth:ble:xxxx:authenticated
- bluetooth:ble:xxxx:connection-control
- bluetooth:ble:xxxx:connected-adapter
- bluetooth:ble:xxxx:rssi
- bluetooth:ble:xxxx:tx-power
- bluetooth:ble:xxxx:estimated-distance
- bluetooth:ble:xxxx:adapter
- bluetooth:ble:xxxx:location
- bluetooth:ble:xxxx:last-updated

**Comment by vkolotov** [comment]
how about "unknown" channels? 

![image](https://user-images.githubusercontent.com/1161883/49123793-d604b000-f31e-11e8-8924-41442e5fde09.png)

Thanks how you do stuff with BT binding. The binding should discover some channels (BT characteristics) automatically. So that you can use them. I assume there will be several channels per each button. If it is like that, you only need to extend the binding with some XML files to get the channels to be recognised automatically, I assume they will be radioboxes/switches in OH.

See here: https://github.com/sputnikdev/eclipse-smarthome-bluetooth-binding/blob/master/gatt-extensions.md

**Comment by vkolotov** [comment]
Once you've changed "GATT parsing strategy" you will need to wait until the binding reloads and then refresh PaperUI page with your things to see new channels. IMPORTANT: they will appear only after the binding receives data, e.g. you will need to switch on/off your buttons to discover all unknown channels.

**Comment by vkolotov** [comment]
This would be an example of "unknown" characteristics (MiFlora device):

![image](https://user-images.githubusercontent.com/1161883/49124195-5aa3fe00-f320-11e8-837e-e4f67c2e55e7.png)


`000071fe-0000-1000-8000-00805f9b34fb` is an unknown characteristic.

**Comment by vkolotov** [comment]
How many buttons does your switch have?

**Comment by vkolotov** [comment]
Right, I've skimmed the doc you have attached.

The way how your switch transmit data is via so called "manufacturer data". The binding supports it, a new channel (single) should be created for it. That channel contains the following (as per the doc):

> Sequence Counter (4 byte)
The Sequence Counter is a continuously incrementing counter used for security
processing. It is initialized to 0 at the time of production and incremented for each
telegram (data telegram or commissioning telegram) sent.
 Switch Status (1 byte)
The Switch Status field reports the button action. The encoding of this field is described
in chapter 4.7.
 Optional Data (0 / 1 / 2 or 4 byte)
PTM 215B provides the option to transmit additional user-defined data within each
data telegram. This data can be used to identify user-specific properties.
The length of the Optional Data field is defined in the Configuration register as described
in chapter 6.7.7.
 Security Signature (4 byte)
The Security Signature is used to authenticate PTM 215B radio telegrams as described
in chapter 4.8

we are interested in `Switch Status (1 byte)`.

**Comment by vkolotov** [comment]
I'll need to double check if the binding supports the "manufacturer data", should be.

**Comment by pfink** [comment]
I have rockers with 2 buttons and also those with 4 buttons. But I think there is the very same radio module behind them. Unfortunately, no additional channels get discovered. I guess this could be because the binding does not know the security key. I think the whole security and commissioning process described in the docs would make no sense if everyone can just publicly read the data ;-)

**Comment by pfink** [comment]
(see section  4.8 in the docs)

**Comment by vkolotov** [comment]
Right, I've checked it :) It is not implementing manufacturer data thingy, but it would be easy to add as this is very-very similar to an existing functionality (advertised service data).

Check this out: https://github.com/sputnikdev/eclipse-smarthome-bluetooth-binding/blob/3bf8b0fe6ebeae386cd55e143fc4c9fdd6748aa6/src/main/java/org/sputnikdev/esh/binding/bluetooth/handler/BluetoothDeviceHandler.java#L157

That one implements "advertised service data" which is very similar to "manufacturer service data", in fact the later should be easier as there will be only 1 data point comparing to many for "service data".

Here is what needs to be implemented:  https://github.com/sputnikdev/bluetooth-manager/blob/d239d59a5368d05d06f59dc62faeb3e355716f23/src/main/java/org/sputnikdev/bluetooth/manager/BluetoothSmartDeviceListener.java#L70

Let me know if it makes sense for you?

**Comment by vkolotov** [comment]
that argument `serviceData ` will contain exactly this:

>Sequence Counter (4 byte)
The Sequence Counter is a continuously incrementing counter used for security
processing. It is initialized to 0 at the time of production and incremented for each
telegram (data telegram or commissioning telegram) sent.
 Switch Status (1 byte)
The Switch Status field reports the button action. The encoding of this field is described
in chapter 4.7.
 Optional Data (0 / 1 / 2 or 4 byte)
PTM 215B provides the option to transmit additional user-defined data within each
data telegram. This data can be used to identify user-specific properties.
The length of the Optional Data field is defined in the Configuration register as described
in chapter 6.7.7.
 Security Signature (4 byte)
The Security Signature is used to authenticate PTM 215B radio telegrams as described
in chapter 4.8

**Comment by vkolotov** [comment]
Where key of that map would be `0x03DA`, and the value is the rest of it:

![image](https://user-images.githubusercontent.com/1161883/49125849-61ce0a80-f326-11e8-9c7c-c4af489c27bf.png)

**Comment by pfink** [comment]
I'll do a detailed review within the next days, thank you so far! In case you have more hints, don't hesitate to append them here :) The longer I read the more I recognize that I probably misunderstood the docs. Now I think that I don't need the security key to read the data, I just need it to verify that it comes from my device and not from an attacker that imitates my device (but for my use cases, I think it's not necessary to protect against such attack scenarios).

On the weekend, I'll setup a proper development environment and try to get it running.

**Comment by vkolotov** [comment]
Yep, the doc does not say that it is absolutely required to implement security checks. For basic usage, it should be more than enough. The telegram message contains signature part, that one can be used to validate if the message is sent by the original device by using keys that was exchanged initially. Not sure if we need it though :), however it is nice to have.

![image](https://user-images.githubusercontent.com/1161883/49126188-8b3b6600-f327-11e8-8658-966eec84fdce.png)

> The receiver performs the same signature calculation based on sequence counter, source
address and the remaining telegram data of the received telegram using the security key it
received from PTM 215B during commissioning.

**Comment by pfink** [comment]
Yes, of course, I also see it definitely as a TODO for the binding. At the very latest when you use BLE devices to open your front door, you should have the posiibility to verify the origin of the telegrams ;-)

**Comment by vkolotov** [comment]
Have a loot at the architecture design of the binding: https://github.com/sputnikdev/eclipse-smarthome-bluetooth-binding/blob/master/implementation-notes.md
This should give you a bit of overview.

**Comment by vkolotov** [comment]
Once you've discovered a new channel for your switch, we will need to think about how could we extend the binding so that it can parse binary telegram message into a set of switches.

Just so you know, at the moment the binding does not have ANY specific logic for any device, everything is made so that it is dynamically creates channels by using BT manager and GATT parser.

I'd like to continue this approach to make the binding as abstract as possible, however looking at the format of telegram message it seems to me it won't be possible. Looks like we need to design a mechanism for the binding that allows us to extend it. First thing that comes to my mind is to use osqi DI to load some custom handlers for unknown/custom characteristics / service data / manufacturer data. For example, once BT device handler comes across an unknown entity, it will try to find a handler in OSGi registry by using a unique ID (in our case if would be UUID `00000059-...` ) and load it if found.

What do you think?

**Comment by pfink** [comment]
> I've added a basic support for advertised manufacturer data into the binding

Thank you very much!

> Could you please grub this snapshot and try your switch with it?

Sure! In this version, one additional String channel named `000003da-0000-0000-0000-000000000000` is discovered. Anyhow, this channel does never send/contain any data.

> Looks like we need to design a mechanism for the binding that allows us to extend it. [..] For example, once BT device handler comes across an unknown entity, it will try to find a handler in OSGi registry by using a unique ID (in our case if would be UUID 00000059-... ) and load it if found.

Absolutely! Sounds very reasonable to me. But I think it would be better not to let the main binding decide upon which criteria things get discovered. Probably it makes more sense that all extensions are subscribed to everything and decide theirselves if something that the BLE binding discovers is relevant for them.

**Comment by pfink** [comment]
Another question: It seems that there is an issue with reliability. With zwave and HomeMatic, I'm used to at least 98% reliability. For the Dolphin/EnOcean PTM 215B, it's 85-95% when the switch is 2.5 meters aways from the adapter. On 8 meters distance, it already drops to 75-85% measured on the real operation as a wall switch. Of course, this is extremely bad for a wall switch.

Under lab/debugging conditions, the reliability is much higher, I'd say > 95%. That's because often the first click after the device was not active for a longer time does not work, while nearly all subsequent clicks are recognized correctly. I observed a very similar behavior when I use [Flic Buttons](https://flic.io/) together with my [Flic Button Binding](https://github.com/pfink/openhab2-flicbutton), so I wouldn't consider this issue related to a specific device. To you have any idea about this or any recommendations to increase reliability? I already tried to:

* Bluez: set controller mode to BLE: `ControllerMode = le`
* Bluez: set `FastConnectable = true`
* Binding: Increase `Bluetooth devices update rate` to 1s.

... but without any success / improvement in reliability, I believe.

**Comment by pfink** [comment]
Last but not least one more thing: With which bluez version would you recommend to setup my development environment? My Raspberry is working with 5.43. I'm tending to set it up based on Ubtuntu Cosmic which has the most recent version (5.50) which has significant changes compared to 5.43 (especially Mesh support).

**Comment by vkolotov** [comment]
Hi @pfink,

Good idea re giving the ability to choose if an extension can do something.

Re reliability, how do you calculate it if you don't even get the data right? :) That new channel, is it empty?

I think I know why you'd get a low reliability rate. Here is it why. I assume you are using a generic transport (not bluegiga transport)? Bluez framework is a terrific piece of junk, furthermore the generic transport talks to Bluez through TinyB library and DBus, this also adds a bit of complexity and less stability. Having said that, there is an issue with Bluez/Dbus/Tinyb (I have not figured out which component is contributing the most) when an object that is responsible for a device gets stale, it just does not report any failure just hangs like there was not any issue. So I have implemented a workaround for this case, it is very basic thing, the BM just checks how long ago was any communication from/to the object and if it exceeds a timeout, it initiates a reset for that object (releases it and gets a new "native" object). Can't remember the timeout, but it should be around 2 mins or so (maybe even less). That's why you probably loose the very first event from the device, because it tries to re-initialize a native object. This is to be confirmed though. 

BTW, in general the blugiga transport is much more stable, however you will still  get that "workaround" mechanism with it. I believe we need to come up with a setting that would control whether that workaround is on/off.

**Comment by vkolotov** [comment]
I'd recommend to switch to BLuegiga :) but I'd also recommend to stay with 4.43 as it proved to be the most stable.

**Comment by pfink** [comment]
Okay, ordered the BLED112 dongle. It's always good for debugging to have the possibility to compare. Then I'll probably setup devenv based on 4.50 to have the newest and hottest shit which will probably not work, and after doing that I can use the BlueGiga for the serious stuff :D

**Comment by vkolotov** [comment]
Is it empty at all? Not values?

**Comment by pfink** [comment]
Yes, no values. Regarding log, I just found these messages:

```
2018-12-02 20:35:29.562 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler BluetoothDeviceHandler of thing bluetooth:ble:E215000053AF tried checking if channel online is linked although the handler was already disposed.
2018-12-02 20:35:29.567 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler BluetoothDeviceHandler of thing bluetooth:ble:E215000053AF tried checking if channel last-updated is linked although the handler was already disposed.
2018-12-02 20:35:29.571 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler BluetoothDeviceHandler of thing bluetooth:ble:E215000053AF tried checking if channel rssi is linked although the handler was already disposed.
2018-12-02 20:35:29.574 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler BluetoothDeviceHandler of thing bluetooth:ble:E215000053AF tried checking if channel tx-power is linked although the handler was already disposed.
2018-12-02 20:35:29.578 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler BluetoothDeviceHandler of thing bluetooth:ble:E215000053AF tried checking if channel estimated-distance is linked although the handler was already disposed.
2018-12-02 20:35:29.581 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler BluetoothDeviceHandler of thing bluetooth:ble:E215000053AF tried checking if channel adapter is linked although the handler was already disposed.
2018-12-02 20:35:29.585 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler BluetoothDeviceHandler of thing bluetooth:ble:E215000053AF tried checking if channel location is linked although the handler was already disposed.
2018-12-02 20:35:29.589 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler BluetoothDeviceHandler of thing bluetooth:ble:E215000053AF tried checking if channel connected is linked although the handler was already disposed.
2018-12-02 20:35:29.592 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler BluetoothDeviceHandler of thing bluetooth:ble:E215000053AF tried checking if channel connection-control is linked although the handler was already disposed.
2018-12-02 20:35:29.596 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler BluetoothDeviceHandler of thing bluetooth:ble:E215000053AF tried checking if channel connected-adapter is linked although the handler was already disposed.
2018-12-02 20:35:29.599 [WARN ] [.core.thing.binding.BaseThingHandler] - Handler BluetoothDeviceHandler of thing bluetooth:ble:E215000053AF tried checking if channel authenticated is linked although the handler was already disposed.
```

**Comment by vkolotov** [comment]
Disregard those errors. It is a different thing that needs to be addressed after OH update.

You will need to enable debug levels for BM module and transport modules.

```
log:set DEBUG org.sputnikdev.bluetooth.manager.impl
log:set DEBUG org.sputnikdev.bluetooth.manager.transport.bluegiga
log:set DEBUG org.sputnikdev.bluetooth.manager.transport.tinyb
```

**Comment by pfink** [comment]
On each click, the following is thrown:

```
2018-12-02 20:54:34.380 [DEBUG] [ooth.manager.impl.DeviceGovernorImpl] - Manufacturer data changed (notification): /00:1A:7D:DA:71:12/E2:15:00:00:53:AF : 1 : 1
2018-12-02 20:54:34.382 [DEBUG] [ager.impl.CombinedDeviceGovernorImpl] - Manufacturer data changed (listener): /XX:XX:XX:XX:XX:XX/E2:15:00:00:53:AF : 1 : true
2018-12-02 20:54:34.613 [DEBUG] [ooth.manager.impl.DeviceGovernorImpl] - Manufacturer data changed (notification): /00:1A:7D:DA:71:12/E2:15:00:00:53:AF : 1 : 1
2018-12-02 20:54:34.615 [DEBUG] [ager.impl.CombinedDeviceGovernorImpl] - Manufacturer data changed (listener): /XX:XX:XX:XX:XX:XX/E2:15:00:00:53:AF : 1 : true
```

**Comment by vkolotov** [comment]
That's good. Bad thing it does not say what changed though :) Things to improve...

**Comment by vkolotov** [comment]
It is still unknown why you don't get anything in PaperUI...

**Comment by pfink** [comment]
Now the following is thrown per click:

```
2018-12-03 21:27:21.005 [TRACE] [.manager.transport.tinyb.TinyBDevice] - RSSI tinyb:/xxxxxxxx/E2:15:00:00:53:xx: -65
2018-12-03 21:27:21.024 [TRACE] [.manager.transport.tinyb.TinyBDevice] - Manufacturer data changed: tinyb:/xxxxxxxx/E2:15:00:00:53:xx: {986=[0f, 02, 00, 00, 05, 44, c7, 27, 7c]}
2018-12-03 21:27:21.156 [TRACE] [.manager.transport.tinyb.TinyBDevice] - RSSI tinyb:/xxxxxxxx/E2:15:00:00:53:xx: -65
2018-12-03 21:27:21.159 [TRACE] [.manager.transport.tinyb.TinyBDevice] - Manufacturer data changed: tinyb:/xxxxxxxx/E2:15:00:00:53:xx: {986=[10, 02, 00, 00, 04, 0e, f4, 01, 57]}
2018-12-03 21:27:21.195 [TRACE] [.manager.transport.tinyb.TinyBDevice] - RSSI tinyb:/xxxxxxxx/E2:15:00:00:53:xx: -65
```

I can also confirm that the data contains the information which button is clicked :)

**Comment by vkolotov** [comment]
Good, do you still get nothing in the PaperUI for that channel?

**Comment by pfink** [comment]
Now something changed... 
![enocean ptm 215b](https://user-images.githubusercontent.com/7902693/49403369-e8cb1a80-f74c-11e8-9fee-1f85c8385ea2.PNG)

**Comment by vkolotov** [comment]
good but it is different Manufacturer ID :) 

0x3da vs 0x986

**Comment by vkolotov** [comment]
Right, it is decimal in logs:
https://www.google.com/search?q=0x3da%20to%20decinal

**Comment by pfink** [comment]
Mhm, but it seems to be a big chance that the item is updated... I cannot reproduce any item updates anymore and on all the other switches it also did not work a single time, anyhow the log message posted above (with changing values, of course) is always printed.

**Comment by vkolotov** [comment]
Are you saying that you get log messages, but it does not reflect to PaperUI?

**Comment by pfink** [comment]
Yes, exactly. I did not observe an update a single time, even though I had plenty of button clicks in the meantime. It still has exactly the same value than in the screenshot, even though I see completely different values in the logs.

**Comment by vkolotov** [comment]
right, strange, this might be a bug.

**Comment by pfink** [comment]
Btw, this debugging/tracing mode is extremely nice! Now I found out why the reliability is so bad: After 20 seconds, it seems like the device gets disposed. After disposal, a wake up needs MUCH more 
signal strength and a higher number of reapeated telegrams than in the "active" mode.

Log:
```
--- Initial wakeup: Needs much more signal strength than normal ---
2018-12-03 21:55:01.112 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Enable RSSI notifications: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.117 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Enable connected notifications: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.119 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Enable service resolved notifications: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.123 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Enable blocked notifications: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.126 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Enable manufacturer data notifications: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.129 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Enable service data notifications: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.132 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Checking if device connected: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.136 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Getting blocked: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.139 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Checking if device connected: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.143 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Getting TxPower: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.146 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Getting bluetooth class: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.149 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Getting ble enabled: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.151 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Getting bluetooth class: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.153 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Getting name: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.155 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Setting alias: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX / E2-15-00-00-53-A2
2018-12-03 21:55:01.158 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Getting RSSI: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.162 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Getting blocked: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.165 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Checking if device connected: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:01.168 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Is services resolved?: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX


--- Repeatedly some Reachability Checks ---
2018-12-03 21:55:21.146 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Getting blocked: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX
2018-12-03 21:55:21.149 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Checking if device connected: tinyb:/XX:XX:XX:XX:XX:XX/XX:XX:XX:XX:XX:XX


--- After 20s, device gets disposed ---
2018-12-03 21:56:17.015 [DEBUG] [.manager.transport.tinyb.TinyBDevice] - Disposing device: XX:XX:XX:XX:XX:XX
```

Is there a way to deactivate this disposal for added BLE things?

**Comment by vkolotov** [comment]
this is what I was talking about, I believe it applies that workaround for stale devices, we will need to add a setting for "thing" to prevent this for this type of devices that goes offline very often but that does not mean that they are offline.

Do you see anything related to "stale" in the logs? There should be something like "checking if object is tale..." (not exactly that).

**Comment by vkolotov** [comment]
Two things needs to be addressed:
* Disabling "stale" checks
* Once a new object is discovered, OH needs to be notified of the current manufacturer data (debatable if it is needed)

**Comment by pfink** [comment]
But I don't really understand why your workaround is necessary at all? Or what does it really work around? From my perspective, it's "by design" that BLE devices only send data if absolutely required. So, by design, of course there can be devices which do not send at least every 20 seconds. Conclusion: The devices do not get stale, they just have to remain active because (by design) you never know when a BLE device will send data again. So you always have to listen.

Wouldn't it also be just possible to let all BLE devices that are added to openHAB as Things remain active forever (until they get removed / disabled by the user)?

**Comment by vkolotov** [comment]
Fund a bug, basically the binding does not subscribe to the event (manufacurer data). This simple. Building now a new snapshot.

**Comment by vkolotov** [comment]
> But I don't really understand why your workaround is necessary at all? Or what does it really work around? From my perspective, it's "by design" that BLE devices only send data if absolutely required.

Believe or not, but your case is an exception, normally BT devices are always online and advertise something. As I mentioned above, TinyB (I suspect) native objects gets stale, irrespectively of a type of the device. Hence there is a workaround.

**Comment by vkolotov** [comment]
Please grub this: https://oss.sonatype.org/content/repositories/snapshots/org/sputnikdev/org.eclipse.smarthome.binding.bluetooth/1.1.7-SNAPSHOT/org.eclipse.smarthome.binding.bluetooth-1.1.7-20181203.223729-4.jar

**Comment by vkolotov** [comment]
There is a project started by @xrucka to get rid of TinyB layer and talk to DBus directly: https://github.com/xrucka/bluetooth-manager-dbus. 

I could not find enough time to test it, it was working for me mostly, but there were a couple of issues still... Would be great if we could ditch TinyB...

**Comment by pfink** [comment]
> normally BT devices are always online and advertise

BT devices (without BLE) -> yes
BLE devices -> I'd say, sometimes, sometimes no

> Hence there is a workaround.

But workaround for what? As I said, I think it's not a bug, it's feature! So what would happen if you just don't dispose things as long as they're added to openHAB? What would stop working then?

**Comment by vkolotov** [comment]
The bug is when you get an instance of a BT object from TinyB library, subscribe for some events, it works for some time (you get events) but then for some reason (not always though, it might work for days) it becomes stale and the object no longer notify you (even if the actual device is online). Hence the workaround to check when an "interaction" was last time and if it exceeded a timeout, destroy the object and acquire new one.

**Comment by pfink** [comment]
Ahh, okay. Now I understand :)

**Comment by vkolotov** [comment]
Let me know if the build above fixes the issue, I do not have time to test it now, I'm having a course now till the end of this week.

**Comment by pfink** [comment]
It doesn't fix the issue. It seems that it gets filled once, but then it's not updated anymore.

**Comment by xrucka** [comment]
@pfink @vkolotov I pretty much rewrote it from the scratch, however have not published it yet (there is still a lot of prooving-ground catch-all code involved), I'll push it to the main repo once I get home today.

**Comment by xrucka** [comment]
@pfink : if you want to try it:
xrucka/bluetooth-manager-dbus branch rewrite_all + xrucka/eclipse-smarthome-bluetooth-binding-dbus-transport rewrite_all

Should work with bluez-5.48+, emulating the battery service. It's not however updated against Vlad's most recent changes, so some things might not work out of the box.

**Comment by pfink** [comment]
Tried BlueGiga with the online channel meanwhile. First test results are worse than tinyB in terms of reliability, it seems like it suffers even more from the "first click not recognized" issue. Will gather more data on the weekend.

EDIT: This time, reducing _Bluetooth devices update rate_ seems to have a quite significant effect. Will continue observing...

**Comment by vkolotov** [comment]
I'll try to fins some time this weekend to implement that setting which turns off the workaround for things.

**Comment by pfink** [comment]
@vkolotov: If you find no time, pls tell me, then I can implement it :) But I'd highly recommend you to set a license for this project before, otherwise you'll run into a legal mess that you can't get out without asking all contributors for permission...

**Comment by pfink** [comment]
FTR: I'm currently testing with BlueGiga and an own build where I completely deactivated the reset of stale devices. First results are quite good. Is there a reason why the "stale devices" workaround is enabled by default for the BlueGiga adapter as well? Doesn't it generally effect reliability negatively, e.g. if telegrams arrive exactly at the moment when the device is reset?

**Comment by vkolotov** [comment]
Hi @pfink, the reason is simple. Transport implementations are supposed to be as simple as possible, focusing on providing an abstract layer between BM and hardware, hence less threads/logics. I'd suggest to keep that logic there in BM for now and implement a setting for device governors so that we can switch it off (a setting per bt device in OH). 

I'm hoping that when we Dbus transport is done and stable, we will get rid of that logic completely.

**Comment by vkolotov** [comment]
> But I'd highly recommend you to set a license for this project before, otherwise you'll run into a legal mess that you can't get out without asking all contributors for permission...

what would be your recommendations? All bits and pieces are licensed with Apache v2.

**Comment by pfink** [comment]
> what would be your recommendations?

As licensing is a quite complex topic, I carved out the answer to #74 :)

**Comment by pfink** [comment]
> I'm hoping that when we Dbus transport is done and stable, we will get rid of that logic completely.

Okay. I'll start testing bluetooth-manager-dbus as soon as I completed the reliability test of my current setup. This will take probably take at least one more week, because I have to test under everday life conditions.

**Comment by vkolotov** [comment]
Great, thank you, let us know how it goes.

**Comment by pfink** [comment]
Reliability of BlueGiga with deactivated stale device removals is very good :) Out of 60 clicks in the last days, just a single one failed (with online channel).

@vkolotov: Anyhow, ManufacturerData still do not arrive at the item.
@xrucka: Tested with dbus, but the buttons do not get discovered. Log output:

```
2018-12-21 22:01:25.709 [INFO ] [.manager.transport.dbus.BluezFactory] - Found running bluetooth daemon, connecting & populating...
2018-12-21 22:01:30.709 [DEBUG] [.manager.transport.dbus.BluezFactory] - Discovered devices:
2018-12-21 22:01:30.709 [DEBUG] [.manager.transport.dbus.BluezFactory] - Discovered adapters:
```
(repeats every few seconds)

I tested it with Bluez  5.50.

Now I'll start to test tinyB with deactivated stale device removals as well.

**Comment by vkolotov** [comment]
Great news. Keeps up posed please. 

Re manufacture data, I believe there is a bug somewhere. I might have some time next week to work on it, it should be something obvious...

**Comment by pfink** [comment]
Findings regarding manufacturer data so far:

- [This code](https://github.com/sputnikdev/eclipse-smarthome-bluetooth-binding/commit/6417a583c5fd166472422dc4abe42f7f6c454dc0) is not called at all
- [This function](https://github.com/sputnikdev/eclipse-smarthome-bluetooth-binding/blob/master/src/main/java/org/sputnikdev/esh/binding/bluetooth/handler/BluetoothDeviceHandler.java#L175) is called when channel is created

**Comment by pfink** [comment]
More findings:

- With BlueGiga transport binding, the buttons get discovered as generic bluetooth devices, not as BLE devices. To get it to work, they have to be added manually with thing type `ble`
- Issue does not always occur. When doing it in the following order, it works fine:
  1. Remove Thing (if present)
  1. Restart openHAB
  1. Add Thing
  1. Press buttons to get the channel added
  1. Link item to a channel
- Anyhow, if this order is not respected, often it does not work. Issue can be safely reproduced by
  1. Add Thing
  1. Press buttons to get the channel added
  1. Restart openHAB
  1. Link item to a channel (can also be done before the restart, it does not matter)
- Technically, what happens is the following:
  - ServiceHandler only listens to changes if the channel is attached to the item
  - [This ](https://github.com/sputnikdev/eclipse-smarthome-bluetooth-binding/blob/master/src/main/java/org/sputnikdev/esh/binding/bluetooth/handler/BluetoothHandler.java#L93) is the method which activates listening to events. It uses the private property [channelHandlers](https://github.com/sputnikdev/eclipse-smarthome-bluetooth-binding/blob/master/src/main/java/org/sputnikdev/esh/binding/bluetooth/handler/BluetoothHandler.java#L45). But, e.g. after openHAB restart, this variable does not contain the ServiceHandler for the dynamic channels. It just contains the [static channels](https://github.com/sputnikdev/eclipse-smarthome-bluetooth-binding/blob/master/src/main/java/org/sputnikdev/esh/binding/bluetooth/handler/GenericBluetoothDeviceHandler.java#L111) which get initialized during instantiation. So, the method tries to attach the ServiceHandler, but it fails because it cannot find it.
  - This error does always occur if the channel is already present when it gets discovered first. This is because in other cases [GattChannelHandler](https://github.com/sputnikdev/eclipse-smarthome-bluetooth-binding/blob/master/src/main/java/org/sputnikdev/esh/binding/bluetooth/handler/GattChannelHandler#L174) does the channel registration (but it skips it, when the channel is already present).

**Comment by pfink** [comment]
I did setup PR #75 now which fixes the issue. Anyhow, I'm not sure if this is a proper fix or just a workaround. @vkolotov: Would be cool if you could take a look on it :)

**Comment by pfink** [comment]
Good news: With the bugfix and some Binary parsing with Jython / JSR223, I'm now able to obtain which button was pressed / released 👍

**Comment by pfink** [comment]
FTR: Textual Configuration works like this.

```
Thing bluetooth:ble:E215000053B8 "My EnOcean PTM 215B Button" {
    Channels:
        Type characteristic--------readable-000003da-0000-0000-0000-000000000000-null: 3DA-3DA "Binary Button Data"
}
```

**Comment by pfink** [comment]
FTR, Textual Configuration works like this:

```
Thing bluetooth:ble:E215000053B1 "My EnOcean PTM 215B Button" {
    Channels:
        Type characteristic--------readable-000003da-0000-0000-0000-000000000000-null: 3DA-3DA "Binary Button Data"
}
```

**Comment by pfink** [comment]
FTR, this is a Python / [JSR223 ](https://www.openhab.org/docs/configuration/jsr223.html) script which parses the binary data and fills other items with human-readable String's that tell what actually happened:

```
from org.slf4j import LoggerFactory
import re

scriptExtension.importPreset("RuleSupport")
scriptExtension.importPreset("RuleSimple")

class MyRule(SimpleRule):
    def __init__(self, rawItemName, targetItemName):
        self.rawItemName = rawItemName
        self.targetItemName = targetItemName

        self.triggers = [
            TriggerBuilder.create()
                    .withId(rawItemName + "-changed")
                    .withTypeUID("core.ItemStateUpdateTrigger")
                    .withConfiguration(
                        Configuration({
                            "itemName": self.rawItemName
                        })).build()
        ]
        
    def execute(self, module, input):
        state = str(ir.getItem(self.rawItemName).getState())
        num_of_bits = 8
        switch_status_hex = state.replace("[", "").replace("]", "").split(", ")[4]
        switch_status_bin = bin(int(switch_status_hex, 16))[2:].zfill(num_of_bits)
        buttons = {
            3: "B1",
            4: "B0",
            5: "A1",
            6: "A0",
        }

        for button in buttons.keys():
            if switch_status_bin[button] == "1":
                button_activated = buttons[button]                

        operation = "PRESSED" if switch_status_bin[7] == "1" else "RELEASED"

        LoggerFactory.getLogger("de.pfink.PTM215BParser").debug("Button Operation: " + button_activated + " " + operation)
        events.postUpdate(self.targetItemName, button_activated + " " + operation)


items = ir.getItem("PTM215B_Button_Events").getAllMembers()

for item in items:    
    targetItemName = item.getName()
    rawItemName = targetItemName + "_Raw"
    automationManager.addRule(MyRule(rawItemName, targetItemName))
```

Sample Item Configuration:

```
Group PTM215B_Button_Events
String Button_Bedroom_Raw  "Button_Bedroom_Raw [%s]"    { channel="bluetooth:ble:E215000053C5:3DA-3DA" }
String Button_Bedroom     "Button_Bedroom [%s]"    (PTM215B_Button_Events)
```

This is how it works: For all items that are member of group `PTM215B_Button_Events`,it tries to find a corresponding item with the same name and the suffix `_Raw`. Then, a rule to that item is attached and listens for changes. If the rule is trigerred, it writes human-readable data into the item which is member of `PTM215B_Button_Events`.

**Comment by pfink** [comment]
FTR, this is a Python / [JSR223 ](https://www.openhab.org/docs/configuration/jsr223.html) script which parses the binary data and fills other items with human-readable String's that tell what actually happened:

```
from org.slf4j import LoggerFactory
import re

scriptExtension.importPreset("RuleSupport")
scriptExtension.importPreset("RuleSimple")

class MyRule(SimpleRule):
    def __init__(self, rawItemName, targetItemName):
        self.rawItemName = rawItemName
        self.targetItemName = targetItemName

        self.triggers = [
            TriggerBuilder.create()
                    .withId(rawItemName + "-changed")
                    .withTypeUID("core.ItemStateUpdateTrigger")
                    .withConfiguration(
                        Configuration({
                            "itemName": self.rawItemName
                        })).build()
        ]
        
    def execute(self, module, input):
        state = str(ir.getItem(self.rawItemName).getState())
        num_of_bits = 8
        switch_status_hex = state.replace("[", "").replace("]", "").split(", ")[4]
        switch_status_bin = bin(int(switch_status_hex, 16))[2:].zfill(num_of_bits)
        buttons = {
            3: "B1",
            4: "B0",
            5: "A1",
            6: "A0",
        }

        for button in buttons.keys():
            if switch_status_bin[button] == "1":
                button_activated = buttons[button]                

        operation = "PRESSED" if switch_status_bin[7] == "1" else "RELEASED"

        LoggerFactory.getLogger("de.pfink.PTM215BParser").debug("Button Operation: " + button_activated + " " + operation)
        events.postUpdate(self.targetItemName, button_activated + " " + operation)


items = ir.getItem("PTM215B_Button_Events").getAllMembers()

for item in items:    
    targetItemName = item.getName()
    rawItemName = targetItemName + "_Raw"
    automationManager.addRule(MyRule(rawItemName, targetItemName))
```

Sample Item Configuration:

```
Group PTM215B_Button_Events
String Button_Bedroom_Raw  "Button_Bedroom_Raw [%s]"    { channel="bluetooth:ble:E215000053D1:3DA-3DA" }
String Button_Bedroom     "Button_Bedroom [%s]"    (PTM215B_Button_Events)
```

This is how it works: For all items that are member of group `PTM215B_Button_Events`,it tries to find a corresponding item with the same name and the suffix `_Raw`. Then, a rule to that item is attached and listens to the incoming binary data. If the rule is trigerred, it parses the data and writes human-readable strings into the item which is member of `PTM215B_Button_Events`.
